File: Python/omniPy/AdvDB/std_write_SAS.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys, os
import pandas as pd
from omniPy.AdvOp import get_values, ExpandSignature
from omniPy.AdvDB import writeSASdat

#[ASSUMPTION]
#[1] If you need to chain the expansion, make sure either of below designs is set
#    [1] Each of the nodes is in a separate module
#    [2] The named instances (e.g. <eSig> here) have unique names among all nodes, if they are in the same module

@(eSig := ExpandSignature(writeSASdat))
def std_write_SAS(
    indat : dict[str, [str | pd.DataFrame]]
    ,outfile : str | os.PathLike
    ,funcConv : callable = lambda x: x
    ,inDat : pd.DataFrame = None
    ,outFile : str | os.PathLike = None
    ,*pos
    ,**kw
) -> int:
    #000. Info.
    '''
#---------------------------------------------------------------------------------------------------------------------------------------#
#100.   Introduction.                                                                                                                   #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |This function acts as a [helper] one to standardize the writing of files or data frames with different processing arguments        #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |CONCLUSION                                                                                                                         #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |[1] Retain the signatures of all ancestor functions for easy program design                                                        #
#   |[2] Add aliases of necessary arguments to enable standardized call                                                                 #
#   |[3] In most cases when this function is called in a standardized way, the first argument is provided in positional pattern         #
#   |[4] In order to call the function with standardized keywords, one must also wrap it with <AdvOp.withDefaults>, see examples        #
#---------------------------------------------------------------------------------------------------------------------------------------#
#200.   Glossary.                                                                                                                       #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |100.   Parameters.                                                                                                                 #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |indat         :   1-item <dict> with its value as the data frame or its literal name (as character string) to be exported, while   #
#   |                   the key of it is not validated, since SAS dataset only contains one data frame per file.                        #
#   |                   [IMPORTANT   ] This argument is for standardization purpose to construct a unified API                          #
#   |outfile       :   PathLike object indicating the full path of the exported data file                                               #
#   |funcConv      :   Callable to mutate the input data frame before exporting it                                                      #
#   |                   [<see def.>  ] <Default> Do not apply further process upon the data                                             #
#   |                   [callable    ]           Callable that takes only one positional argument with data.frame type                  #
#   |inDat         :   The same argument in the ancestor function, which is a placeholder in this one, omitted and overwritten as       #
#   |                   <indat> is of different input type so it no longer takes effect                                                 #
#   |                   [IMPORTANT] We always have to define such argument if it is also in the ancestor function, and if we need to    #
#   |                   supersede it by another argument. This is because we do not know the <kind> of it in the ancestor and that it   #
#   |                   may be POSITIONAL_ONLY and prepend all other arguments in the expanded signature, in which case it takes the    #
#   |                   highest priority during the parameter input. We can solve this problem by defining a shared argument in this    #
#   |                   function with lower priority (i.e. to the right side of its superseding argument) and just do not use it in the #
#   |                   function body; then inject the fabricated one to the parameters passed to the call of the ancestor.             #
#   |                   [<see def.>  ] <Default> Use the same input as indicated in <indat>                                             #
#   |outFile       :   The same argument in the ancestor function, which is a placeholder in this one, superseded by <outfile> so it no #
#   |                   longer takes effect                                                                                             #
#   |                   [<see def.>  ] <Default> Use the same input as <outfile>                                                        #
#   |*pos          :   Various positional arguments to expand from its ancestor; see its official document                              #
#   |**kw          :   Various keyword arguments to expand from its ancestor; see its official document                                 #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |900.   Return Values by position.                                                                                                  #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |<int>         :   Return code from the encapsulated function call                                                                  #
#---------------------------------------------------------------------------------------------------------------------------------------#
#300.   Update log.                                                                                                                     #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   | Date |    20240101        | Version | 1.00        | Updater/Creator | Lu Robin Bin                                                #
#   |______|____________________|_________|_____________|_________________|_____________________________________________________________#
#   | Log  |Version 1.                                                                                                                  #
#   |______|____________________________________________________________________________________________________________________________#
#   |___________________________________________________________________________________________________________________________________#
#   | Date |    20240129        | Version | 1.10        | Updater/Creator | Lu Robin Bin                                                #
#   |______|____________________|_________|_____________|_________________|_____________________________________________________________#
#   | Log  |[1] Remove the unnecessary restrictions on arguments, and leave them to the caller process                                  #
#   |______|____________________________________________________________________________________________________________________________#
#   |___________________________________________________________________________________________________________________________________#
#   | Date |    20250201        | Version | 2.00        | Updater/Creator | Lu Robin Bin                                                #
#   |______|____________________|_________|_____________|_________________|_____________________________________________________________#
#   | Log  |[1] Introduce <ExpandSignature> to expand the signature with those of the ancestor functions for easy program design        #
#   |______|____________________________________________________________________________________________________________________________#
#---------------------------------------------------------------------------------------------------------------------------------------#
#400.   User Manual.                                                                                                                    #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |See the [Full Test Program] section                                                                                                #
#---------------------------------------------------------------------------------------------------------------------------------------#
#500.   Dependent Facilities.                                                                                                           #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |100.   Dependent Modules                                                                                                           #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |   |sys, os, pandas                                                                                                                #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |300.   Dependent user-defined functions                                                                                            #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |   |AdvDB                                                                                                                          #
#   |   |   |writeSASdat                                                                                                                #
#   |   |-------------------------------------------------------------------------------------------------------------------------------#
#   |   |AdvOp                                                                                                                          #
#   |   |   |get_values                                                                                                                 #
#   |   |   |ExpandSignature                                                                                                            #
#---------------------------------------------------------------------------------------------------------------------------------------#
    '''

    #010. Check parameters.
    #011. Prepare log text.
    #python 动态获取当前运行的类名和函数名的方法: https://www.cnblogs.com/paranoia/p/6196859.html
    LfuncName : str = sys._getframe().f_code.co_name

    #012. Handle the parameter buffer.
    if not isinstance(indat, dict):
        raise TypeError(f'[{LfuncName}]<indat> must be a 1-item dict, while <{type(indat).__name__}> is given!')
    if len(indat) != 1:
        raise ValueError(f'[{LfuncName}]<indat> must be a 1-item dict, while <{len(indat)}> items are given!')

    #013. Define the local environment.

    #100. Identify the shared arguments between this function and its ancestor functions
    val_in = list(indat.values())[0]
    if isinstance(val_in, str): val_in = get_values(val_in, inplace = False, instance = pd.DataFrame)
    val_out = funcConv(val_in)
    args_share = {
        'inDat' : val_out
        ,'outFile' : outfile
    }
    eSig.vfyConflict(args_share)

    #700. Insert the patched values into the input parameters
    pos_out, kw_out = eSig.insParams(args_share, pos, kw)

    #999. Return the result
    return( eSig.src(*pos_out, **kw_out) )
#End std_write_SAS

'''
#-Notes- -Begin-
#Full Test Program[1]:
if __name__=='__main__':
    #010.   Create environment.
    import sys
    import os
    import pandas as pd
    dir_omniPy : str = r'D:\Python\ '.strip()
    if dir_omniPy not in sys.path:
        sys.path.append( dir_omniPy )
    from omniPy.AdvOp import withDefaults
    from omniPy.AdvDB import std_write_SAS
    from omniPy.Dates import asDates, asDatetimes, asTimes

    #100. Load the meta config table for data conversion
    conv_meta = pd.read_excel(os.path.join(dir_omniPy, 'omniPy', 'AdvDB', 'meta_writeSASdat.xlsx'))

    #200. Create data frame in terms of the indication in above meta config table
    aaa = (
        pd.DataFrame(
            {
                'var_str' : 'abcde'
                ,'var_int' : 5
                ,'var_float' : 14.678
                ,'var_date' : '2023-12-25'
                ,'var_dt' : '2023-12-25 12:34:56.789012'
                ,'var_time' : '12:34:56.789012'
                ,'var_ts' : asDatetimes('2023-12-25 12:34:56.789012', fmt = '%Y-%m-%d %H:%M:%S.%f')
            }
            ,index = [0]
        )
        #Prevent pandas from inferring dtypes of these fields
        .assign(**{
            'var_date' : lambda x: asDates(x['var_date'])
            #<%f> is only valid at input (strptime) rather than output (strftime)
            ,'var_dt' : lambda x: asDatetimes(x['var_dt'], fmt = '%Y-%m-%d %H:%M:%S.%f')
            ,'var_time' : lambda x: asTimes(x['var_time'], fmt = '%H:%M:%S.%f')
        })
    )

    #300. Convert the data to SAS dataset
    outf = os.path.join(os.getcwd(), 'vfysas.sas7bdat')
    rc = std_write_SAS(
        {'key' : aaa}
        ,outf
        ,metaVar = conv_meta
    )
    if os.path.isfile(outf): os.remove(outf)

    #500. Standardize the call with keyword input pattern
    #[ASSUMPTION]
    #[1] <withDefaults> only enables keyword provision of parameters, while any positional or keyword-only arguments without
    #     default values should still be provided as input in accordance with the signature
    std_param = {
        'indat' : {'key' : aaa}
        ,'encoding' : 'GB2312'
        ,'outfile' : outf
    }

    #510. Wrap the function to enable provision of keyword parameters
    writesas = withDefaults(std_write_SAS)
    help(writesas)

    #550. Write data
    rc2 = writesas(**std_param)
    if os.path.isfile(outf): os.remove(outf)

#-Notes- -End-
'''

File: R/omniR/AdvDB/std_write_SAS.r
#---------------------------------------------------------------------------------------------------------------------------------------#
#100.   Introduction.                                                                                                                   #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |This function acts as a [helper] one to standardize the writing of files or data frames with different processing arguments        #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |Scenarios:                                                                                                                         #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |[1] We could pass various parameters into one single expression [kw] that have no negative impact to current function call         #
#---------------------------------------------------------------------------------------------------------------------------------------#
#200.   Glossary.                                                                                                                       #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |100.   Parameters.                                                                                                                 #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |indat         :   1-item <list> with its value as the data frame or its literal name (as character string) to be exported, while   #
#   |                   the key of it is not validated, since SAS dataset only contains one data frame per file.                        #
#   |                   [IMPORTANT   ] This argument is for standardization purpose to construct a unified API                          #
#   |outfile       :   PathLike object indicating the full path of the exported data file                                               #
#   |funcConv      :   Function to mutate the input data frame before exporting it                                                      #
#   |                   [<see def.>  ] <Default> Do not apply further process upon the data                                             #
#   |                   [function    ]           Function that takes only one positional argument with data.frame type                  #
#   |inDat         :   The same argument in the ancestor function, which is a placeholder in this one, omitted and overwritten as       #
#   |                   <indat> is of different input type so it no longer takes effect                                                 #
#   |                   [IMPORTANT] We always have to define such argument if it is also in the ancestor function, and if we need to    #
#   |                   supersede it by another argument. This is because we do not know the <kind> of it in the ancestor and that it   #
#   |                   may be POSITIONAL_ONLY and prepend all other arguments in the expanded signature, in which case it takes the    #
#   |                   highest priority during the parameter input. We can solve this problem by defining a shared argument in this    #
#   |                   function with lower priority (i.e. to the right side of its superseding argument) and just do not use it in the #
#   |                   function body; then inject the fabricated one to the parameters passed to the call of the ancestor.             #
#   |                   [<see def.>  ] <Default> Use the same input as indicated in <indat>                                             #
#   |outFile       :   The same argument in the ancestor function, which is a placeholder in this one, superseded by <outfile> so it no #
#   |                   longer takes effect                                                                                             #
#   |                   [<see def.>  ] <Default> Use the same input as <outfile>                                                        #
#   |...           :   Various named parameters for the encapsulated function call if applicable                                        #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |900.   Return Values by position.                                                                                                  #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |<int>         :   Return code from the encapsulated function call                                                                  #
#---------------------------------------------------------------------------------------------------------------------------------------#
#300.   Update log.                                                                                                                     #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   | Date |    20240215        | Version | 1.00        | Updater/Creator | Lu Robin Bin                                                #
#   |______|____________________|_________|_____________|_________________|_____________________________________________________________#
#   | Log  |Version 1.                                                                                                                  #
#   |______|____________________________________________________________________________________________________________________________#
#   |___________________________________________________________________________________________________________________________________#
#   | Date |    20250214        | Version | 2.00        | Updater/Creator | Lu Robin Bin                                                #
#   |______|____________________|_________|_____________|_________________|_____________________________________________________________#
#   | Log  |[1] Introduce <ExpandSignature> to expand the signature with those of the ancestor functions for easy program design        #
#   |______|____________________________________________________________________________________________________________________________#
#---------------------------------------------------------------------------------------------------------------------------------------#
#400.   User Manual.                                                                                                                    #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |See the [Full Test Program] section                                                                                                #
#---------------------------------------------------------------------------------------------------------------------------------------#
#500.   Dependent Facilities.                                                                                                           #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |100.   Dependent Modules                                                                                                           #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |   |rlang, glue                                                                                                                    #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |300.   Dependent user-defined functions                                                                                            #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |   |AdvDB                                                                                                                          #
#   |   |   |writeSASdat                                                                                                                #
#   |   |-------------------------------------------------------------------------------------------------------------------------------#
#   |   |AdvOp                                                                                                                          #
#   |   |   |isVEC                                                                                                                      #
#   |   |   |get_values                                                                                                                 #
#   |   |   |ExpandSignature                                                                                                            #
#---------------------------------------------------------------------------------------------------------------------------------------#

#001. Append the list of required packages to the global environment
#Below expression is used for easy copy-paste from raw text strings instead of quoted ones.
lst_pkg <- deparse(substitute(c(
	rlang, glue
)))
#Quote: https://www.regular-expressions.info/posixbrackets.html?wlr=1
lst_pkg <- paste0(lst_pkg, collapse = '')
lst_pkg <- gsub('[[:space:]]', '', lst_pkg, perl = T)
lst_pkg <- gsub('^c\\((.+)\\)', '\\1', lst_pkg, perl = T)
lst_pkg <- unlist(strsplit(lst_pkg, ',', perl = T))
options( omniR.req.pkg = base::union(getOption('omniR.req.pkg'), lst_pkg) )

#We should use the big-bang operand [!!!] supported by below package
library(rlang)

#[ASSUMPTION]
#[1] When we need the immediate effect of the signature expansion, we need to instantiate <ExpandSignature> before the function
#     is defined
#[2] To make the instance holding various methods unique, we set both the instance and the function in the same <local> environment
#[3] Using <local()> will immediately evaluate all statements inside it, hence we have to prepare sufficient resources before
#     executing below script, e.g. ensure <ExpandSignature> is defined in current environment
std_write_SAS <- local({
#[ASSUMPTION]
#[1] By instantiation of below class, we resemble a <class decorator> in Python
deco <- ExpandSignature$new(writeSASdat, instance = 'eSig')
myfunc <- deco$wrap(function(
	indat
	,outfile
	,funcConv = function(x) x
	,inDat = NULL
	,outFile = NULL
	,...
){
	#010. Parameters
	#[Quote: https://stackoverflow.com/questions/15595478/how-to-get-the-name-of-the-calling-function-inside-the-called-routine ]
	LfuncName <- deparse(sys.call()[[1]])
	#If above statement cannot find the name correctly, this function must have been called via [do.call] or else,
	# hence we need to traverse one layer above current one and extract the first argument of that call.
	if (grepl('^function.+$',LfuncName[[1]],perl = T)) LfuncName <- gsub('^.+?\\((.+?),.+$','\\1',deparse(sys.call(-1)),perl = T)[[1]]

	#012. Handle the parameter buffer.
	if (!all(typeof(indat) == 'list')) {
		stop(glue::glue('[{LfuncName}]<indat> must be a 1-item list, while <{toString(typeof(indat))}> is given!'))
	}
	if (length(indat) != 1) {
		stop(glue::glue('[{LfuncName}]<indat> must be a 1-item list, while <{toString(length(indat))}> are given!'))
	}

	#020. Local environment.
	dots <- rlang::list2(...)

	#100. Identify the shared arguments between this function and its ancestor functions
	val_in <- indat[[1]]
	if (isVEC(val_in)) if (is.character(val_in)) {
		val_in <- get_values(val_in, inplace = F, mode = 'list')
	}
	val_out <- funcConv(val_in)
	args_share <- list(
		'inDat' = val_out
		,'outFile' = outfile
	)
	eSig$vfyConflict(args_share)

	#700. Insert the patched values into the input parameters
	args_out <- eSig$updParams(args_share, dots)

	#999. Return the result
	return(do.call(eSig$src, args_out))
})
return(myfunc)
})

#[Full Test Program;]
if (FALSE){
	#Simple test
	if (TRUE){
		#010. Load user defined functions
		source('D:\\R\\autoexec.r')
		if (nchar(Sys.getenv('TZ')) == 0) Sys.setenv(TZ = 'Asia/Shanghai')

		library(magrittr)

		#500. Convert the data to SAS dataset without meta config table
		#[ASSUMPTION]
		#[1] Dtypes that are not involved below CANNOT be exported, and will lead to exceptions
		testdf <- data.frame(
			var_str = c('abcde',NA)
			,var_raw = c(as.raw(40), charToRaw('A'))
			,var_int = c(5,7)
			,var_float = c(14.678,83.32)
			,var_date = c('2023-12-25','2023-12-32')
			,var_dt = c('2023-12-25 12:34:56.789012','2023-12-31 00:24:41.16812')
			,var_time = c('12:34:56.789012','789')
			,var_bool = c(T,F)
			,var_cat = as.factor(c('abc','def'))
			,var_complex = c(1 + 3i, 12.4 + 4.6i)
			,stringsAsFactors = F
		) %>%
			dplyr::mutate(
				var_int = as.integer(var_int)
				,var_date = asDates(var_date)
				,var_dt = asDatetimes(var_dt)
				,var_time = asTimes(var_time)
			)

		outf2 <- file.path(getwd(), 'vfysas2.sas7bdat')
		rc <- std_write_SAS(
			list('vfy' = testdf)
			,outf2
		)
		if (file.exists(outf2)) rc <- file.remove(outf2)

	}
}
